File: Data-Structure-I/98-Validate-Binary-Search-Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool isValidBST(TreeNode* root) {
        return check(root, LONG_MIN, LONG_MAX);
    }
    bool check(TreeNode* node, long int min, long int max){
        if(node==nullptr) return true;
        if(!(node->val>min && node->val<max)) return false;
        return check(node->left, min, node->val) && check(node->right, node->val, max);
    }
};

File: Data-Structure-I/98-Validate-Binary-Search-Tree.php
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     public $val = null;
 *     public $left = null;
 *     public $right = null;
 *     function __construct($val = 0, $left = null, $right = null) {
 *         $this->val = $val;
 *         $this->left = $left;
 *         $this->right = $right;
 *     }
 * }
 */
class Solution {

    /**
     * @param TreeNode $root
     * @return Boolean
     */
    function isValidBST($root) {
        return $this->check($root, PHP_INT_MIN, PHP_INT_MAX);
    }
    
    function check($node, $min, $max){
        if($node==NULL) return true;
        if(!($node->val<$max && $node->val>$min)) return false;
        return $this->check($node->left, $min, $node->val) && $this->check($node->right, $node->val, $max);
    }
}
